include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -246,6 +252,90 @@ struct __parallel_for_submitter<__internal::__optional_kernel_name<_Name...>>
     }
 };
 
+template <typename _KernelName, typename... _Ranges>
+struct __parallel_for_large_submitter;
+
+template <typename... _Name, typename... _Ranges>
+struct __parallel_for_large_submitter<__internal::__optional_kernel_name<_Name...>, _Ranges...>
+{
+    // Flatten the range as std::tuple value types in the range are likely coming from separate ranges in a zip
+    // iterator.
+    using _FlattenedRangesTuple = typename oneapi::dpl::__internal::__flatten_std_or_internal_tuple<
+        std::tuple<oneapi::dpl::__internal::__value_t<_Ranges>...>>::type;
+    using _MinValueType = typename oneapi::dpl::__internal::__min_tuple_type<_FlattenedRangesTuple>::type;
+    // __iters_per_work_item is set to 1, 2, 4, 8, or 16 depending on the smallest type in the
+    // flattened ranges. This allows us to launch enough work per item to saturate the device's memory
+    // bandwidth. This heuristic errs on the side of launching more work per item than what is needed to
+    // achieve full bandwidth utilization. 16 bytes per range per work item has been found as a good
+    // value across the different for-based algorithms.
+    static constexpr std::uint8_t __bytes_per_work_item = 16;
```

Comment:
Since the reduce discussions, I believe there have been some improvements to coalesce style accesses in the compiler. For some cases with parallel for, there is benefit for adding a vertically vectorized path (sequential processing within a work item) but getting the compiler to generate the correct instructions is tricky to do for all cases.

For 32-bit and 64-bit types, I have seen similar performance with `std::copy` and a handcrafted version which generates vectorized loads / stores. Interestingly, there seems to be small benefits for `std::for_each` for these types. For smaller types such as 16-bit, vectorization brings significant improvement as the compiler is able to combine memory transactions. The bulk of further performance improvements would come with these small types I believe.

The reason that I have not added it now is that it has been non-trivial to implement in a generic fashion with our `__parallel_for` pattern to accommodate the different bricks. I have seen spotty performance thus far when trying to do so. When doing just a vector load such as in reduce, the compiler seems to easily generate the correct instructions. For load / store such as in `std::copy` it has been difficult to enable vectorization with our bricks. If the compiler doesn't generate the correct vector instructions, the performance significantly degrades. I think it is possible to implement a version that does this, but requires additional work that should come in a subsequent round of improvements that build on top of these. For the most common cases, this PR closes most of the gap and puts us near the memory bandwidth roof.

Reply:
Thanks for so detailed reply. I agree: no need to do it in this PR.
 If you think that would be useful and feasible, I suggest leaving a TODO comment.

Reply:
There is a TODO comment in the kernel submitter about adding a vectorized path.